packages/daimo-api/src/contract/paymaster.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  DEFAULT_USEROP_PREVERIFICATION_GAS_LIMIT,
+  DEFAULT_USEROP_CALL_GAS_LIMIT,
+  DEFAULT_USEROP_VERIFICATION_GAS_LIMIT,
+} from "@daimo/common";
+import { tokenMetadata } from "@daimo/contract";
+import { concat, pad, parseAbi, toHex } from "viem";
+
+import { ViemClient } from "../chain";
+
+const pimlicoPaymasterAbi = parseAbi([
+  "function priceMarkup() view returns (uint32)",
+  "function previousPrice() view returns (uint192)",
+]);
+
+/* Interface to on-chain paymaster and gas related data. */
+export class Paymaster {
+  /* Gas price constants */
+  private maxFeePerGas: bigint = 0n;
+  private maxPriorityFeePerGas: bigint = 0n;
+
+  /* Paymaster constants */
+  private priceMarkup: bigint = 0n;
+  private previousPrice: bigint = 0n;
+
+  private lastFetchDT: Date = new Date(0);
+
+  constructor(private vc: ViemClient) {}
+
+  async init() {
+    this.fetchLatestState();
+  }
+
+  private async fetchLatestState() {
+    // Do not fetch more than once per 5 minute for now.
+    // todo: Ask Pimlico to change contract to emit events on updates
+    // todo: fetch fees seperately on higher frequency
+    if (new Date().getTime() - this.lastFetchDT.getTime() < 5 * 60 * 1000)
+      return;
+
+    const priceMarkup = await this.vc.publicClient.readContract({
+      abi: pimlicoPaymasterAbi,
+      address: tokenMetadata.paymasterAddress,
+      functionName: "priceMarkup",
+    });
+    const previousPrice = await this.vc.publicClient.readContract({
+      abi: pimlicoPaymasterAbi,
+      address: tokenMetadata.paymasterAddress,
+      functionName: "previousPrice",
+    });
+
+    const maxFeePerGas = await this.vc.publicClient.getGasPrice();
+    const maxPriorityFeePerGas =
+      await this.vc.publicClient.estimateMaxPriorityFeePerGas(); // Assumes we're on an EIP-1559 chain
+
+    this.lastFetchDT = new Date();
+    this.priceMarkup = BigInt(priceMarkup);
+    this.previousPrice = previousPrice;
+    this.maxFeePerGas = maxFeePerGas;
+    this.maxPriorityFeePerGas = maxPriorityFeePerGas;
+  }
+
+  // Since our various gas limits corresponding to the userop are fixed, we can
+  // compute these constants as a pure function.
+  // Leftover gas payment is refunded by the paymaster so overpaying is fine.
```

Comment:
amazing

(also, all of our userops cost a similar amount, w signature verification as the dominant term)